server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java

```diff
@@ -35,12 +42,48 @@
 
     private final ShardId shardId;
     private final FilterDirectoryReader.SubReaderWrapper wrapper;
+    private final CheckedFunction<DirectoryReader, String, IOException> readerIdGenerator;
+    private final String readerId;
 
-    private ElasticsearchDirectoryReader(DirectoryReader in, FilterDirectoryReader.SubReaderWrapper wrapper,
-            ShardId shardId) throws IOException {
+    private ElasticsearchDirectoryReader(DirectoryReader in, FilterDirectoryReader.SubReaderWrapper wrapper, ShardId shardId,
+                                         CheckedFunction<DirectoryReader, String, IOException> readerIdGenerator) throws IOException {
         super(in, wrapper);
         this.wrapper = wrapper;
         this.shardId = shardId;
+        this.readerIdGenerator = readerIdGenerator;
+        this.readerId = readerIdGenerator.apply(in);
+    }
+
+    /**
+     * If two readers have the same reader id, then their underlying reader must consist of the same list of segments.
+     */
+    @Nullable
+    public String getReaderId() {
+        return readerId;
+    }
+
+    public static String getReaderIdFromSegmentInfos(SegmentInfos segmentInfos, BooleanSupplier fullyCommitted) {
```

Comment:
I am not sure if we should pursue this. But without it, we'd never have the same searcher id between copies except with searchable snapshots.